File: examples/reference-services/python/client_server_llm/prompter.py
#!/usr/bin/env python3
"""
Prompter for the Python reference demo (CONTROL-only).

Single command usage:
  python prompter.py --file plan_prompt.txt
  python prompter.py --file run_trigger.txt

Behavior:
- If the file parses as JSON and looks like a scheduler CONTROL run
  (has {"to":"scheduler", "stage":"run", ...}), it is sent as CONTROL
  with content type application/vnd.sw4rm.scheduler.command+json;v=1.
- Otherwise, the file content is wrapped as CONTROL stage=prompt to the
  scheduler with the same content type.
"""
import json
import os
import sys
import grpc

# Add SDK to path when running from this directory
sys.path.append(os.path.join(os.path.dirname(__file__), "..", "..", "..", "sdks", "py_sdk"))

from sw4rm.clients.router import RouterClient
from sw4rm.envelope import build_envelope
from sw4rm import constants as C


def main() -> int:
    import argparse
    ap = argparse.ArgumentParser()
    ap.add_argument("--file", required=True, help="Path to the plan or run file")
    args = ap.parse_args()

    file_path = args.file
    if not os.path.isfile(file_path):
        print(f"file not found: {file_path}")
        return 2

    router_host = os.getenv("ROUTER_HOST", "localhost")
    router_port = int(os.getenv("ROUTER_PORT", "50051"))
    channel = grpc.insecure_channel(f"{router_host}:{router_port}")
    router = RouterClient(channel)

    raw = open(file_path, "r", encoding="utf-8").read()
    try:
        obj = json.loads(raw)
    except Exception:
        obj = None

    if isinstance(obj, dict) and (obj.get("to") in (None, "scheduler")) and obj.get("stage") == "run":
        payload = raw.encode("utf-8")
        env = build_envelope(
            producer_id="prompter",
            message_type=C.CONTROL,
            content_type="application/vnd.sw4rm.scheduler.command+json;v=1",
            payload=payload,
        )
    else:
        payload = json.dumps({
            "schema_version": 1,
            "to": "scheduler",
            "stage": "prompt",
            "params": {"prompt": raw},
        }).encode("utf-8")
        env = build_envelope(
            producer_id="prompter",
            message_type=C.CONTROL,
            content_type="application/vnd.sw4rm.scheduler.command+json;v=1",
            payload=payload,
        )

    try:
        resp = router.send_message(env)
        accepted = getattr(resp, "accepted", False)
        reason = getattr(resp, "reason", "")
        print("accepted=", accepted, "reason=", reason)
        return 0 if accepted else 1
    except Exception as e:
        print("send failed:", e)
        return 1


if __name__ == "__main__":
    raise SystemExit(main())

File: sdks/js_sdk/reference-services/agents/prompter.ts
#!/usr/bin/env tsx
// moved to agents/

// Prompter for JS demo: send CONTROL seed/run messages to scheduler.

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import yargs from 'yargs/yargs';
import { hideBin } from 'yargs/helpers';

import { RouterClient, buildEnvelope, MessageType } from '../../../../sdks/js_sdk/src/index.ts';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

async function main() {
  const argv = await yargs(hideBin(process.argv))
    .option('file', { type: 'string', demandOption: true, describe: 'Path to plan or run file' })
    .strict()
    .parse();

  const filePath = path.resolve(String(argv.file));
  if (!fs.existsSync(filePath)) {
    console.error('file not found:', filePath);
    process.exit(2);
  }

  const ROUTER_HOST = process.env.ROUTER_HOST || 'localhost';
  const ROUTER_PORT = Number(process.env.ROUTER_PORT || '50051');
  const router = new RouterClient({ address: `${ROUTER_HOST}:${ROUTER_PORT}` });

  const raw = fs.readFileSync(filePath, 'utf8');
  let obj: any | null = null;
  try { obj = JSON.parse(raw); } catch {}

  let payload: Buffer;
  if (obj && typeof obj === 'object' && (obj.to === undefined || obj.to === 'scheduler') && obj.stage === 'run') {
    payload = Buffer.from(raw, 'utf8');
  } else {
    payload = Buffer.from(JSON.stringify({
      schema_version: 1,
      to: 'scheduler',
      stage: 'prompt',
      params: { prompt: raw },
    }), 'utf8');
  }

  const env = buildEnvelope({
    producer_id: 'prompter',
    message_type: MessageType.CONTROL,
    content_type: 'application/vnd.sw4rm.scheduler.command+json;v=1',
    payload,
  });
  try {
    const res = await router.sendMessage(env);
    console.log('accepted=', res.accepted, 'reason=', res.reason || '');
    process.exit(res.accepted ? 0 : 1);
  } catch (e: any) {
    console.error('send failed:', e?.message || e);
    process.exit(1);
  }
}

main();
